packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js

```diff
@@ -0,0 +1,309 @@
+import useOnBlockDrop from '../use-on-block-drop';
+
+/** @typedef {import('../../utils/math').WPPoint} WPPoint */
+
+/**
+ * The type of a drag event.
+ *
+ * @typedef {'default'|'file'|'html'} WPDragEventType
+ */
+
+/**
+ * An array representing data for blocks in the DOM used by drag and drop.
+ *
+ * @typedef {Object} WPBlockNavigationDropZoneBlocks
+ * @property {string}  clientId                        The client id for the block.
+ * @property {string}  rootClientId                    The root client id for the block.
+ * @property {number}  blockIndex                      The block's index.
+ * @property {Element} element                         The DOM element representing the block.
+ * @property {number}  innerBlockCount                 The number of inner blocks the block has.
+ * @property {boolean} isDraggedBlock                  Whether the block is currently being dragged.
+ * @property {boolean} canInsertDraggedBlocksAsSibling Whether the dragged block can be a sibling of this block.
+ * @property {boolean} canInsertDraggedBlocksAsChild   Whether the dragged block can be a child of this block.
+ */
+
+/**
+ * An object containing details of a drop target.
+ *
+ * @typedef {Object} WPBlockNavigationDropZoneTarget
+ * @property {string}                   blockIndex   The insertion index.
+ * @property {string}                   rootClientId The root client id for the block.
+ * @property {string|undefined}         clientId     The client id for the block.
+ * @property {'top'|'bottom'|'inside'}  dropPosition The position relative to the block that the user is dropping to.
+ *                                                   'inside' refers to nesting as an inner block.
+ */
+
+/**
+ * A react hook that returns data about blocks used for computing where a user
+ * can drop to when dragging and dropping blocks.
+ *
+ * @param {Object}          ref           A React ref of a containing element for block navigation.
+ * @param {WPPoint}         position      The current drag position.
+ * @param {WPDragEventType} dragEventType The drag event type.
+ *
+ * @return {WPBlockNavigationDropZoneBlocks} An array representing data for each block in the block navigation DOM.
+ */
+function useDropTargetBlocksData( ref, position, dragEventType ) {
+	const {
+		getBlockRootClientId,
+		getBlockIndex,
+		getBlockCount,
+		getDraggedBlockClientIds,
+		canInsertBlocks,
+	} = useSelect( ( select ) => {
+		const {
+			canInsertBlocks: _canInsertBlocks,
+			getBlockRootClientId: _getBlockRootClientId,
+			getBlockIndex: _getBlockIndex,
+			getBlockCount: _getBlockCount,
+			getDraggedBlockClientIds: _getDraggedBlockClientIds,
+		} = select( 'core/block-editor' );
+		return {
+			canInsertBlocks: _canInsertBlocks,
+			getBlockRootClientId: _getBlockRootClientId,
+			getBlockIndex: _getBlockIndex,
+			getBlockCount: _getBlockCount,
+			getDraggedBlockClientIds: _getDraggedBlockClientIds,
+		};
```

Comment:
Nitpick: I don't think this is necessary, couldn't we just assign it to a variable and assign them inline in the return statement?

```suggestion
		const selector = select( 'core/block-editor' );
		return {
			canInsertBlocks: selector.canInsertBlocks,
			getBlockRootClientId: selector.getBlockRootClientId,
			getBlockIndex: selector.getBlockIndex,
			getBlockCount: selector.getBlockCount,
			getDraggedBlockClientIds: selector.getDraggedBlockClientIds,
		};
```

Reply:
This is something that's becoming more an more prevalent. I like that pattern you propose. done in 9d9e228b00c2996aeb8370801725c6343e1fce4e